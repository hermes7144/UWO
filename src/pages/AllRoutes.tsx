import React from 'react';
import { useQuery } from '@tanstack/react-query';
import { getRoutes } from '../api/firebaseTest';
import RouteCard from '../components/RouteCard';
import { useNavigate } from 'react-router-dom';
import Button from '../components/ui/Button';

type RouteType = {
  id: string;
  title: string;
  remark: string;
  citys: number[];
};

export default function AllRoutes() {
  const { isLoading, error, data: routes } = useQuery(['routes'], getRoutes);
  const navigate = useNavigate();

  const handleClick = () => {
    navigate('/routes/new');
  };
  return (
    <section>
      <div className='flex justify-end m-2'>
        <Button text={'새 무역경로'} onClick={handleClick}></Button>
      </div>

      {isLoading && <p>Loading...</p>}
      {error && <p>Error</p>}
      <ul>{routes && routes.map((route: RouteType) => <RouteCard key={route.id} route={route} />)}</ul>
    </section>
  );
}
