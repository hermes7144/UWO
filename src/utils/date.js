// import { format, register } from 'timeago.js';
// import koLocale from 'timeago.js/lib/lang/ko';

// register('ko', koLocale);

// export function formatAgo(date, lang = 'en_US') {
//   return format(date, lang);
// }

export function formatDate(rawDate) {
  const date = new Date(rawDate);
  const year = date.getFullYear().toString().slice(-2); // 연도에서 뒤의 두 자리만 추출
  const month = ('0' + (date.getMonth() + 1)).slice(-2); // 월을 두 자리 숫자로 변환
  const day = ('0' + date.getDate()).slice(-2); // 일을 두 자리 숫자로 변환
  return `${year}/${month}/${day}`;
}