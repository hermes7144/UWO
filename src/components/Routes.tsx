import { useNavigate } from 'react-router-dom';
import { useRouteHooksContext } from '../context/RouteHooksContext';
import { formatDate } from '../utils/date';

type RouteType = {
  id: string;
  title: string;
  description: string;
  createdAt: number;
  citys: number[];
  country: string;
  region: string;
  startMonth: number;
  endMonth: number;
};

export default function Routes() {
  const { useRoute } = useRouteHooksContext();
  const navigate = useNavigate();

  const {
    routesQuery: { isLoading, error, data: routes },
  } = useRoute();

  return (
    <>
      {isLoading && <p>Loading...</p>}
      {error && <p>Something is wrong</p>}
      {routes && (
        <table className='min-w-full text-center text-sm font-light'>
          <colgroup>
            <col width={'5%'} />
            <col width={'5%'} />
            <col width={'5%'} />
            <col width={'50%'} />
            <col width={'10%'} />
            <col width={'10%'} />
          </colgroup>
          <thead className='border-b bg-neutral-50 font-medium dark:border-neutral-500 dark:text-neutral-800'>
            <tr>
              <th scope='col' className='px-1 py-4'>
                국가
              </th>
              <th scope='col' className='px-1 py-4'>
                지역
              </th>
              <th scope='col' className='px-1 py-4'>
                기간
              </th>
              <th scope='col' className='px-1 py-4'>
                제목
              </th>
              <th scope='col' className='px-1 py-4'>
                작성일
              </th>
              <th scope='col' className='px-1 py-4'>
                추천
              </th>
            </tr>
          </thead>
          <tbody>
            {routes.map((route: RouteType) => (
              <tr key={route.id} className='border-b dark:border-neutral-500'>
                <td className='whitespace-nowrap px-1 py-2'>{route.country}</td>
                <td className='whitespace-nowrap px-1 py-2'>{route.region}</td>
                <td className='whitespace-nowrap px-1 py-2'>
                  {route.startMonth} ~ {route.endMonth}월
                </td>
                <td className=' overflow-hidden px-2 py-4 text-left text-ellipsis whitespace-nowrap' style={{ maxWidth: '250px' }}>
                  <span
                    className='cursor-pointer hover:underline'
                    onClick={() => {
                      navigate(`/routes/${route.id}`);
                    }}>
                    {route.title}
                  </span>
                </td>
                <td className='whitespace-nowrap px-2 py-2'>{formatDate(route.createdAt)}</td>
                <td className='whitespace-nowrap  px-2 py-2'>0</td>
              </tr>
            ))}
          </tbody>
        </table>
      )}
    </>
  );
}
