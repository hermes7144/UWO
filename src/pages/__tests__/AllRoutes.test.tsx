import { Route } from 'react-router-dom';
import AllRoutes from '../AllRoutes';
import { render, screen } from '@testing-library/react';
import { withAllContexts, withRouter } from '../../tests/utils';
import { fakeRoute } from '../../tests/fakeRoutes';
import Routes from '../../components/Routes';

jest.mock('../../components/Routes');

describe('AllRoutes', () => {
  const fakeUseRoute = jest.fn();
  let user = null;

  beforeEach(() => {
    fakeUseRoute.mockImplementation(() => {
      return {
        routeQuery: { isLoading: false, data: fakeRoute },
      };
    });
  });

  afterEach(() => {
    fakeUseRoute.mockReset();
    Routes.mockReset();
    user = null;
  });

  it('renders correctly', () => {
    const { asFragment } = renderAllRoutes();
    expect(asFragment()).toMatchSnapshot();
  });

  it('user logined and show new button', () => {
    user = { id: 'test' };
    renderAllRoutes();
    expect(screen.getByText('New')).toBeInTheDocument();
  });

  function renderAllRoutes() {
    return render(withAllContexts(withRouter(<Route path='/' element={<AllRoutes />} />), fakeUseRoute, user));
  }
});
