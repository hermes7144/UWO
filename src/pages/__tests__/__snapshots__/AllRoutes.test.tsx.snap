// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllRoutes renders correctly 1`] = `
<DocumentFragment>
  <div
    class="p-2"
  >
    <div>
      <div
        class="flex flex-col lg:flex-row items-center justify-center p-2"
      >
        <label
          class="text-brand font-bold block min-w-fit"
          for="country"
        >
          국가:
        </label>
        <select
          class="p-1 m-1 w-full max-w-xs border-2 outline-none mr-5"
          id="country"
        >
          <option>
            공통
          </option>
          <option>
            포트투갈
          </option>
          <option>
            에스파냐
          </option>
          <option>
            잉글랜드
          </option>
          <option>
            네덜란드
          </option>
          <option>
            오스만
          </option>
          <option>
            명
          </option>
          <option>
            조선
          </option>
          <option>
            일본
          </option>
        </select>
        <label
          class="text-brand font-bold block min-w-fit"
          for="region"
        >
          주요 지역:
        </label>
        <select
          class="p-1 m-1 w-full max-w-xs border-2 outline-none mr-5"
          id="region"
        >
          <option>
            전체
          </option>
          <option>
            동지중해
          </option>
          <option>
            서지중해
          </option>
          <option>
            북해
          </option>
          <option>
            카리브해
          </option>
          <option>
            서아프리카
          </option>
          <option>
            동아프리카
          </option>
          <option>
            아랍
          </option>
          <option>
            서인도
          </option>
          <option>
            동인도
          </option>
          <option>
            서남아
          </option>
          <option>
            동남아
          </option>
          <option>
            동아시아
          </option>
          <option>
            기타
          </option>
        </select>
        <label
          class="text-brand font-bold block min-w-fit"
          for="month"
        >
          월:
        </label>
        <select
          class="p-1 m-1 w-full max-w-xs border-2 outline-none"
          id="month"
        >
          <option>
            전체
          </option>
          <option>
            1
          </option>
          <option>
            2
          </option>
          <option>
            3
          </option>
          <option>
            4
          </option>
          <option>
            5
          </option>
          <option>
            6
          </option>
          <option>
            7
          </option>
          <option>
            8
          </option>
          <option>
            9
          </option>
          <option>
            10
          </option>
          <option>
            11
          </option>
          <option>
            12
          </option>
        </select>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
